### PR TITLE
MF-1525 - Add graceful stop for HTTP and GRPC servers

### LINKED ISSUE #1525: Consider using Go native error groups in main functions
**ENHANCEMENT**
1. Describe the enhancement you are requesting.
Consider using Go errors group (`golang.org/x/sync/errgroup`) for handling routines in `main` functions of Mainflux services.

2. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
This is a nice-to-have.

### PR DESCRIPTION
**Pull request by arvindh123**
### What does this do?
This pull request adds a graceful shutdown to HTTP and gRPC servers.

### Which issue(s) does this PR fix/relate to?
This pull request resolves #1525.

### List any changes that modify/break current functionality
There are no such changes.

### Have you included tests for your changes?
Yes.

### Did you document any new/modified functionality?
No, there is no need for that.

### Notes
N/A

### CODE REVIEW

File: cmd/auth/main.go
Comment (drasko) [inline cmd/auth/main.go]: Is this `else` really needed?

**Review by drasko — CHANGES_REQUESTED** [review]

File: cmd/auth/main.go
Comment (arvindh123) [reply cmd/auth/main.go]: yes, its required, since the function doesn't return in if condition and need to wait for ctx.Done() channel after the if else ,

 Alternatively we can use switch with case true and default 
 Example
```go
switch certFile != "" || keyFile != "" {
case true : 
	logger.Info(fmt.Sprintf("Authentication service started using https, cert %s key %s, exposed port %s", certFile, keyFile, port))
	go func() {
		errs <- server.ListenAndServeTLS(certFile, keyFile)
		return
	}()
default : 
	logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
	go func() {
		errs <- server.ListenAndServe()
		return
	}()
}

```

Comment (dborovcanin) [reply cmd/auth/main.go]: This looks better. Even a switch with no condition is OK:
```go
switch {
	case certFile != "" || keyFile != "":
		logger.Info(fmt.Sprintf("Authentication service started using https, cert %s key %s, exposed port %s", certFile, keyFile, port))
		go func() {
			errs <- server.ListenAndServeTLS(certFile, keyFile)
		}()
	default:
		logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
		go func() {
			errs <- server.ListenAndServe()
		}()
	}
	<-ctx.Done()
	if err := server.Shutdown(context.Background()); err != nil {
		errs <- fmt.Errorf("server shutdown failed:%v", err)
	}
```
Please note that `return` is removed since it is the last statement in the function. Also, notice `fmt.Errorf("server shutdown failed:%v", err)` because errors should not be capitalized and there is no need for `%+v`, but just `%v`.

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: No need for `return`.

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: Same here.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: Please use context with the deadline here to prevent infinite waiting for shutdown:
```suggestion
c, done := context.WithTimeout(ctx, time.Second)
defer done()
if err := server.Shutdown(c); err != nil {
	logger.Error(fmt.Sprintf("Authentication %s service error occured during shutdown at %s: %s", proto, p, err))
	return fmt.Errorf("Authentication %s service error occured during shutdown at %s: %w", proto, p, err)
}
logger.Info(fmt.Sprintf("authentication %s service shutdown of http at %s", proto, p))
return nil
```

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: Please don't change the default log level.

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: Please rename, `proto` is usually associated with `protobuf`.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: cmd/auth/main.go
Comment (arvindh123) [reply cmd/auth/main.go]: Shall we have env variable to set timeout duration or let it be constant for 1 or x  second

Comment (dborovcanin) [reply cmd/auth/main.go]: I think constant (~1s) is fine. We don't have to configure everything, since the user already has a lot of env vars to take care of. I also prefer constant over magic numbers.

File: pkg/errors/errors.go
Comment (dborovcanin) [inline pkg/errors/errors.go]: Consider renaming to `SignalHandler` since this actually does not handle kill signal. KILL signal cannot be trapped. Maybe we should also add `SIGABRT` here, as well.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please resolve conflicts (you can wait for #1358 to be merged first; we'll do it soon).

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: Fix typo `occurred`. Same for all the same typos in other services.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
This looks good. Once the typo is fixed, it will be ready.

File: cmd/coap/main.go
Comment (dborovcanin) [inline cmd/coap/main.go:237]: This is not HTTP, this is CoAP (or underlying UDP).

File: cmd/mqtt/main.go
Comment (dborovcanin) [inline cmd/mqtt/main.go]: This is not `for HTTP` (also, please use `HTTP` instead of `http`), but for `MQTT`.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .vscode/settings.json
Comment (manuio) [inline .vscode/settings.json]: Remove this file please

**Review by manuio — CHANGES_REQUESTED** [review]

File: .vscode/settings.json
Comment (drasko) [reply .vscode/settings.json]: You can also add `.vscode` files to `.gitignore`

File: cmd/auth/main.go
Comment (drasko) [inline cmd/auth/main.go]: Consider renaming variable. From current naming it looks like we are waiting to get samo grace, not waiting for a stop. Maybe `stopWaitTime` or `stopTimeout`? Also, take care of the letter case (`Time`, not `TIme`).

**Review by drasko — CHANGES_REQUESTED** [review]

File: cmd/auth/main.go
Comment (dborovcanin) [inline cmd/auth/main.go]: `graceWaitTime` constant can already be calculated like `5*time.Second` so that you use it like `time.After(graceWaitTime)`. I agree with @drasko, maybe simple `stopTimeout` is a better naming.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Except for one minor remark, this looks good and ready to merge.

File: .vscode/settings.json
Comment (arvindh123) [reply .vscode/settings.json]: done

File: .gitignore
Comment (dborovcanin) [inline .gitignore]: Please add this to **your** global `.gitignore`. While VSCode is defacto the default editor for Go, we don't want to keep its config in the project `.gitignore` file (it is not files generated by Mainflux or for the Mainflux project needs).

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Other than `.gitignore` remark, this is good.
Once we merge this one, we should start thinking about #1557 to simplify HTTP and gRPC servers' configuration.

**Review by dborovcanin — DISMISSED** [review]

**Comment by manuio** [comment]
@arvindh123 Please, update your branch and resolve the conflicts

File: .gitignore
Comment (manuio) [inline .gitignore]: new line at the end of the file

Comment (arvindh123) [reply .gitignore]: Added new line to end of the file

**Review by dborovcanin — DISMISSED** [review]

**Review by manuio — DISMISSED** [review]
LGTM

File: cmd/bootstrap/main.go
Comment (drasko) [inline cmd/bootstrap/main.go]: `Shutdown` is one word, so no need for `ShutDown`

File: cmd/bootstrap/main.go
Comment (drasko) [inline cmd/bootstrap/main.go]: Move string constant to const

**Review by drasko — CHANGES_REQUESTED** [review]

File: cmd/bootstrap/main.go
Comment (arvindh123) [reply cmd/bootstrap/main.go]: During the service start , the log will happen with the  service name and protocol in which service is getting started , 

Purpose of the variable is , 
During graceful shutdown, Need to log the service  with exact protocol ,
So initialized a variable `protocol` with `http` value 

If the service start as http , then variable `protocol` will be same as initialized   `http` , i.e no change in variable 

else if the service started as https , then the variable  `protocol` is changed to `https`

As said above , Need to log the service with exact protocol during graceful shutdown, so on looking the logs ,it will clear view able with service name  with protocol,

Comment (manuio) [reply cmd/bootstrap/main.go:336]: You can move both as constants

**Review by dborovcanin — DISMISSED** [review]

**Review by manuio — DISMISSED** [review]
LGTM

File: cmd/auth/main.go
Comment (drasko) [inline cmd/auth/main.go]: Please move string constants at the top, in the `const` section

**Review by drasko — CHANGES_REQUESTED** [review]

File: cmd/auth/main.go
Comment (arvindh123) [reply cmd/auth/main.go]: Hi @drasko ,



The variable protocol is reuse to find the service is started as HTTP or HTTPS 

By default variable `protocol` is set to `http` 
https://github.com/arvindh123/mainflux/blob/8accfe2c8dff57ca143c0924bfe0d872e60dd6e1/cmd/auth/main.go#L270



If the service start as HTTPS , https://github.com/arvindh123/mainflux/blob/8accfe2c8dff57ca143c0924bfe0d872e60dd6e1/cmd/auth/main.go#L277
then the  variable `protocol` is set to `https`


The detailed explanation is given in the [https://github.com/mainflux/mainflux/pull/1548#discussion_r835408692](https://github.com/mainflux/mainflux/pull/1548#discussion_r835408692
)

Comment (arvindh123) [reply cmd/auth/main.go:270]: Hi @drasko ,



The variable protocol is reuse to find the service is started as HTTP or HTTPS 

By default variable `protocol` is set to `http` 
https://github.com/arvindh123/mainflux/blob/8accfe2c8dff57ca143c0924bfe0d872e60dd6e1/cmd/auth/main.go#L270



If the service start as HTTPS , https://github.com/arvindh123/mainflux/blob/8accfe2c8dff57ca143c0924bfe0d872e60dd6e1/cmd/auth/main.go#L277
then the  variable `protocol` is set to `https`


The detailed explanation is given in the [https://github.com/mainflux/mainflux/pull/1548#discussion_r835408692](https://github.com/mainflux/mainflux/pull/1548#discussion_r835408692
)

Comment (manuio) [reply cmd/auth/main.go]: @arvindh123 He means to do for example:
```
const (
   httpProto  = "http"
   httpsProto = "https"
)
```
And tthen reuse this constants.

Comment (arvindh123) [reply cmd/auth/main.go]: @manuio , I'm asking out of eager to know, Why we need to use in this way? .

Comment (manuio) [reply cmd/auth/main.go]: Because it's consistent with the rest of the code. You can also add a nice name to your constants to make them understandable.

Comment (arvindh123) [reply cmd/auth/main.go]: yeah I agree, but I had thought, why need to waste space in for a variable, I know it was stupid, 
. as I still comparing GO with Assembly Language of 8085

Comment (dborovcanin) [reply cmd/auth/main.go]: I think this is acceptable without constants and even more readable this way.

**Review by manuio — DISMISSED** [review]
LGTM

**Comment by drasko** [comment]
@arvindh123 please update the branch.

@dusanb94 can we please push this one forward.

**Comment by dborovcanin** [comment]
> @arvindh123 please update the branch.
> 
> @dusanb94 can we please push this one forward.

As soon as @arvindh123 updates the PR, it's ready.

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**
